polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -999,38 +1005,50 @@ LoopInternalization::getDimsThatUseLoopIV(LoopLikeOpInterface loop,
 
 Value LoopInternalization::getVersionCondition(
     LoopLikeOpInterface loop, const WorkGroupSize &workGroupSize,
-    DataFlowSolver &solver) const {
+    std::variant<Value, unsigned> reqdSharedMemory,
+    const unsigned sharedMemoryRemaining, DataFlowSolver &solver) const {
+  OpBuilder builder(loop);
+  Value versionCond;
+
+  // Check for shared memory availability.
+  if (std::holds_alternative<Value>(reqdSharedMemory))
+    versionCond = builder.create<arith::CmpIOp>(
+        loop.getLoc(), arith::CmpIPredicate::ule,
+        std::get<Value>(reqdSharedMemory),
+        ValueOrUnsigned::getValue(sharedMemoryRemaining, builder));
+
   // Find out which dimensions use the loop IV in the memory accesses that
   // should use shared memory. For example given (assume 'i' is the loop IV):
   //   A[tx][ty][i], B[tx][i][ty]
   // The set collected would contain {0,1}.
   std::set<unsigned> dimsThatUseLoopIV = getDimsThatUseLoopIV(loop, solver);
   if (dimsThatUseLoopIV.size() <= 1)
-    return nullptr;
+    return versionCond;
 
   // If the memory accesses do not use the loop induction variable
   // 'consistently' (that is on the same dimension) we need to version the
   // loop. Build the versioning condition as:
   //   (wgSize[dim1] == wgSize[dim2]) && wgSize[dim1] == wgSize[dim3] && ...)
-  OpBuilder builder(loop);
-  Value versionCond, first;
+  Value cond, first;
   for (unsigned dim : dimsThatUseLoopIV) {
     Value wgSize = ValueOrUnsigned::getValue(workGroupSize[dim], builder);
 
-    if (!versionCond)
-      versionCond = first = wgSize;
-    else if (isa<arith::CmpIOp, arith::AndIOp>(versionCond.getDefiningOp()))
-      versionCond = builder.create<arith::AndIOp>(
-          loop.getLoc(), versionCond,
+    if (!cond)
+      cond = first = wgSize;
+    else if (isa<arith::CmpIOp, arith::AndIOp>(cond.getDefiningOp()))
+      cond = builder.create<arith::AndIOp>(
+          loop.getLoc(), cond,
           builder.create<arith::CmpIOp>(loop.getLoc(), arith::CmpIPredicate::eq,
                                         first, wgSize));
     else
-      versionCond = builder.create<arith::CmpIOp>(
+      cond = builder.create<arith::CmpIOp>(
           loop.getLoc(), arith::CmpIPredicate::eq, first, wgSize);
   }
-  LLVM_DEBUG(llvm::dbgs() << "versionCond: " << versionCond << "\n");
+  assert(cond && "cond should not be a nullptr");
 
-  return versionCond;
+  return versionCond
+             ? builder.create<arith::AndIOp>(loop.getLoc(), versionCond, cond)
+             : cond;
```

Comment:
I think we can make the loop simpler by hoisting `first` and `cond` initialization outside the loop (use `dimsThatUseLoopIV[0]` for `first` and `dimsThatUseLoopIV[1] == first` for `cond`). That way, the loop will look simpler, as `first` won't be modified and only the body of the `else if` branch will be needed.

Reply:
done